### ISSUE TITLE
"Wait for multiple responses“

### ISSUE THREAD

**Issue by periclesJ** [rapidpro/rapidpro#864]
Hello guys, I'd like to propose a new ruleset called 'wait for multiple responses' this is really useful to handle questions that must have many responses in a same wait for response. For example: 

"Give the numbers of all the kinds of food you like(Separated by comma):"
1. Pizza
2. Hamburger
3. Eggs
4. Bacon
5. French fries
6. Pancakes

Inside the response we could have an array called "categories" (_@flow.response.categories_) that will contain all the options chosen by the user [category_name1, category_name2, category_name5, category_name6]. This way we can access all categories that might be fulfilled by its index (_@flow.response.categories.3_) The ruleset will not have multiple outgoing paths, only a single one, and If it's necessary to handle different situations for each response the user might user a split on the array values.

Sometimes is really necessary to check multiple characteristics in the same response.
![multipleresponses](https://user-images.githubusercontent.com/38072244/52058027-5b862580-2545-11e9-86af-95360f7cfb72.jpg)

**Comment by rowanseymour** [comment]
We do have something like this called "message forms" so that users can more easily parse multiple values from the same input, e.g.

<img width="669" alt="captura de pantalla 2019-01-31 a la s 16 07 41" src="https://user-images.githubusercontent.com/675558/52085392-64d9b700-2572-11e9-953b-1514c95cf788.png">

It's more designed for a scenario where you have a fixed number of values. For example if you had an app that expected inputs like `<cmd> <number> <item>`, e.g. "BUY 30 ORANGES"

**Comment by nicpottier** [comment]
Ya, this is definitely something that would be nice, having multiple answers isn't the most uncommon thing in the world. It does pose some challenges as you pointed out, both around routing and storage of results. We've thought about various approaches there, I think the most logical still being you get routed to the "first matching" rule (but all matching categories are recorded). That would give people some flexibility to route differently if they wanted based on priority of the answer (more complicated routing could be done via expressions perhaps). 

So the routing case isn't so bad really, essentially it is no different.

Where it gets tricky is the storage of results and categories. Values are currently determined by the matching rule, as our categories. So it would involve the introduction of an array type for results. Which I would then say probably means that all results and category matches are arrays, just with the simple case being of length 1. We'd then need to figure out what that means with respect to contact fields, what it means in visualizations of results, exports etc..

In short, it isn't not a small thing. That said, I think it probably is the *right* thing long term, so we have to figure out what our story there is. The trick is to make it not more complicated and weird from a user perspective, so we would need good default behavior in expressions, contact fields etc for array types.

Definitely worth some thought as something to work towards after new engine is out. @ericnewcomer

**Comment by rowanseymour** [comment]
Maybe instead of routing on the category of the first value, you could collapse the exits into "All Categorized" and "Other". That way you have a way to route if any value didn't match a rule, and re-prompt the user.  

For example you ask the contact to enter 3 numbers and they send "1 2 X". The node just has a `has_number` rule so 1 and 2 can be categorized, but because "X" can't, we route to "Other", and then the flow prompts the contact to try again.

**Comment by nicpottier** [comment]
I think I could probably be convinced of either, perhaps it is clearer to users especially from a path count perspective to have all the valid responses rolled up, it would also provide visual indication at the top level of a ruleset being a multi-response ruleset.

Either way the tricky bit is how we store arrays of values / categories though..

**Comment by rowanseymour** [comment]
I think it depends on what the primary need here is (and that's still not completely clear for me). Categorization serves a few different purposes: validation, routing, reporting. If the need is to be able to say that an input consisting of N values is valid because all N values match one of a set of rules, then I think you have to collapse the exits.

If it's validation of number lists then we could avoid complicating results completely by providing a test function like `has_numbers(min, max)`.

@periclesJ can you give a bit more information about a real world flow that would have benefitted from this?

**Comment by nicpottier** [comment]
Oh man, I don't know, that's pretty crazy and something totally nutty. I would much rather we explore the idea of having array fields and then you could have dynamic groups based on the value of that field for example. But running through multiple exits of a ruleset is not something I think we want to do, that totally breaks the flow paradigms of only moving forward and only being on a single path, which are important things to maintain.

**Comment by nicpottier** [comment]
So have been thinking about this in the background and I do want to have an answer for this in the new engine (at least have a plan for it even if we don't launch new engine with it).

The more I've thought about it the more I think that instead of having an array type and then basically results of arrays, instead flip it on its head and say that we have arrays of results, but by default in a flow you are looking at the last result that has been set.

I think this has a few interesting / good properties. 

1. At its simplest, there is value to know all the ways someone responded to a wait, even if the last is by default what we expose to users. For the super typical case you can see how people answered in a way that failed validation before answering correctly, which is interesting data that we don't really expose in any way currently (short of recent messages).

2. At the next level of complexity, our current answer to a flow that wants people to enter more than one thing is to make people start subflows. (IE, a subflow for each member of the family) That is probably still the right answer for cases where it is a complex set of questions that you want to have answered, but you could imagine something more like "Respond with a color you like, or done when complete" and looping back to that wait, appending each value they enter. In the end you have an array of results for that value that has all the colors they like. (plus done)

3. It allows for a fairly logical way to deal with the case above, where you want to specify a bunch of categories and evaluate all of them for a particular message, IE, the "select many" case where you want to allow actions based on each category. You could have a flow just as pictured above, but with a path from each of those "add to group" actions back into the ruleset. This would fit into the current flow model perfectly well and each iteration through that router would append a new result to the array of results for that name.

I think that also highlights that really arrays of results is the only way this makes sense as opposed to having typed arrays. We are loosely typed so each result in an array of results may be of a totally different type.

As to what this would like like, a first shwag may be:

1. Currently, we don't actually expose `results` anywhere. Our runs endpoint: https://textit.in/api/v2/runs thinks in terms of `values` because we didn't want to break the API, but this may be a blessing in disguise. Maybe `values` on a run are the last result recorded and we can augment our runs with `results` which is an array of results:

```json
"results": [{
  "value": "blue",
  "category": "Blue",
  "node": "fc32aeb0-ac3e-42a8-9ea7-10248fdf52a1",
  "time": "2015-11-11T13:03:51.635662Z",
  "name": "color",
  "input": "it is blue"
},
{
  "value": "",
  "category": "Other",
  "node": "fc32aeb0-ac3e-42a8-9ea7-10248fdf52a1",
  "time": "2015-11-11T13:03:51.635662Z",
  "name": "color",
  "input": "I want a unicorn"
}
]
```

2. The use of `values` here begs the question of whether we should introduce `@values` within a flow to mean the same thing. (the last value for a result) That said, I think that may be confusing honestly, would be nicer to figure out how to expose the arrayability of a `@results` value instead. I could probably be convinced either way though.

3. Session and run format would have to change results to be an array per above.

4. Exports could stay as they are for now (exporting `values` not `results`) but we could add the option in the future to export `results` instead?

5. API could just add a `results` attribute to our runs.

I think I really like this as a paradigm that addresses use cases we don't currently have good answers for in a pretty clean way. Thoughts @rowanseymour @ericnewcomer @rowanseymour ?

**Comment by rowanseymour** [comment]
To be honest I'm still not entirely sure what what problem we're trying to solve here, so maybe it would be useful to think about how questions like this currently work and identify the specific deficiencies we want to address.

For example if I want to ask a contact for multiple answers to a single question, I'd probably have an initial ruleset to get the input, validate that the user gave me a set of numbers, and then handle each possible value sequentially by looking at the initial response to see what it contains:

<img width="503" src="https://user-images.githubusercontent.com/675558/54160742-5a86c480-441e-11e9-9282-874f945dff8c.png">

Cons

1. There is some repetition with the rulesets for each possible answer
2. There's no obvious association between the results, e.g. no obvious way for me to later output a list of the foods this person picked.

Pros

1. Simplicity of each answer being as a separate result (`results.likes_pizza`, `results.likes_burgers`), i.e. easy to determine whether the contact likes burgers - I don't have to know that is stored at `results.foods[1]` etc. No ordering to worry about.
2. Flow is linear and I don't have to remember to loop each thing back to a single router.

Can we sketch out what simpler routing for this example would look like? I'm having a hard time picturing how the select many would work with `Other` that would want to wait for new input, but other categories are just looping thru a tokenization of the current input.

**Comment by rowanseymour** [comment]
Also should mention that we do already store the history of a result's value as `run_result_changed` events. Obviously not exposed to expressions but could one day appear in contact history page that renders sessions like sessionexplain.

**Comment by nicpottier** [comment]
I don't think we need to think very hard as to the use case here. Select multiple is a feature of pretty much every polling platform and polling is indeed one of the core things RapidPro is meant to solve. (we also know that we have customers trying to do this, Keith and his regexes were trying to solve this very problem)

So I think this is very much something we need to support and that's why I want to make sure we button this up while we have some flexibility instead of introducing more changes to context etc later.

Yes, you could build a flow as you have. I don't necessarily hate the multiple splits, though that will get pretty tiring, but I do not agree that having results named `likes_burgers`, `likes_pizza` etc is an acceptable user experience. If we did support array results then that same method could be used however. (though we'd have to think through special handling of the "other" case)

Yes, we store all the events, but they aren't results, can't be exported and aren't available to users in any structured way (in any way really, but even if we showed them on the contact page that wouldn't be enough).

Note that FLOIP does support `select_many`: https://github.com/FLOIP/flow-results/blob/master/specification.md#select_many

As do most platforms where you might ship results for visualization etc.. If we have a reasonable answer to how those are represented then people can use those results naturally, having prefixes like `likes_` is not likely to work well in that world.

**Comment by rowanseymour** [comment]
Sure .. am not disagreeing that people aren't trying to ask multiple choice questions. Am trying to understand what part of the current experience we're trying to fix before we rush into a big change to how results are modeled.

I think the whole modeling of results thing is potentially a different question to adding new ways of routing. For example you could have a checkbox option on a ruleset to tokenize the input and apply rules to each token. The exits from that ruleset might be just "All" and "Not All" - where "All" means every token was categorized and "Not all" means not every token matched a category, and you'd naturally loop that back to the same node via a send msg to re-ask the contact for input.

Now if you want to act on each possible value in the answer, you're still going to have to do multiple checks later to see if X is one of the answers. But maybe that's ok. It depends on what problem we're trying to solve. It's certainly got the advantage of simplicity, and..

1. You get validation that the input is a set of valid values.
2. There's no distinction between going back into the node to wait for input vs going back in as part of a loop.

Users like Keith would probably like 1. and that's one of the reasons I feel skeptical about just keeping result history. If I have result history like @result.foods[0...N] that's going to include answers that didn't validate.

Let's say I ask a user for multiple choice answers as numbers, and they first send "1 2 X" and then "X 2 3" and finally "1 2 3".. I don't want to process any results until I get "1 2 3".

( I only mention results being stored as events because I don't see much general utility in storing result history if it doesn't help us provide a better solution to multiple choice )

**Comment by nicpottier** [comment]
Totally agree storage and routing are different things. I think that's why I'm more interested in figuring out what our results storage answer is for multiple choice first, if we pick the right primitive solution there then we can figure out what kind of flow paradigm makes sense to populate it. It would be nice if that solution unlocked other things by being fairly generic.

So to that point, looping back to the original node for multiple evaluations is one way to leverage arrays of results, but I don't think the only way (or even the best, its just an idea) to do so. Could be something handled completely by a new router type for example and we handle extra actions per category with result tests later on in the flow for example. Important first bit is that we have the right primitive types to allow for the natural representations of select many results.

Not sure I understand the approach you are proposing above. Result would just be the string "1 2 3" when it matches? Would that just be a string? Or is that an array type?

I also don't want to overfit Keith's case. I think your first example of `what are your favorite foods` should also match responses like `i like pizza and burgers`. IE, I don't think matching every single token should be a requirement.

You are right that inclusion of invalid values in our results is not a desirable thing in the case where you only want the valid values. We do this today, though that is overwritten if the user does end up responding with something valid so is generally minimized. But ya, that definitely doesn't feel right. Maybe the "default" exit is treated differently?

**Comment by rowanseymour** [comment]
I guess I feel kinda the opposite - I'm not sure we can be confident we've picked the right results format if we can't describe in concrete terms how we would use it. And by concrete I mean really flesh out how something like multiple choice would work. 

In my mind there's 3 parts to multiple choice:

1. Validation (aka Keith's regexes). If you're asking a user to reply with a set of numbers 1-3 you need way to reject responses like "1, 2, 4"
2. Results modeling. What do these results look like in exports? Does our format lend itself to be being consumed in visualization tools or converted to other formats like FLOIP?
3. Simplified routing. This is where this ticket started tho frankly this is the one I'm struggling with the most as it's hard to think of any solution here that doesn't break assumptions about how flows flow.

I would say it's more important to be able to reject `1, 2, 4` than accept `I like 1 and 2 and 3` so I would lean toward stricter validation that says every token has to match a rule. In my mind it should be basic requirement of a multiple choice solution to be able to verify that the input is a set of the possible answers. Otherwise you're forcing users to fallback on regexes etc.

The "Not All / All" router I've described above takes care of 1 and I think would lend itself to modeling results as arrays of values to take care of 2. I.e. you're not looping through something accumulating results - you're saying this set of values is the result, and that set of values _in the aggregate_ are categorizable.

I think the only other solution being proposed is the router which initially waits, then loops through tokens categorizing them one at a time (can we a shorthand name for this?? ). I think we need to sketch this out in a bit more detail because in my head it feels overly complicated. Routing back in from Other has to wait for new input, but routing back in from other categories doesn't. Do all categories have to be connected back to the router? What if there are no actions leading from a category.. you can't route the router to itself etc.

**Comment by nicpottier** [comment]
Agree that 3) is pretty edge-casy. The original example is probably better served with better support for searching on field values / dynamic groups rather than assigning to a manual group.

I would like to support both rejecting `1,2,4` and allowing `i like pizza and beer`. If we find a solution to the latter then I think that unlocks more and makes me more confident in our data model and paradigm. IE, it would be nice if this wasn't some super special case which I feel the "not all/all" is.

So the core of what I was proposing was just that we have arrays of results instead of results of arrays. How we populate that is a different thing. Could be through the looping I initially described which kinda allows 3) but agree has some weirdnesses. Could be a router that just evaluates all switches and creates a result for each matching rule.

I think in a way that is probably very similar to what you are proposing, the only question if whether we force every single token to validate, which I feel is not something we should need to decide from a datamodel perspective. (IE, maybe some users want that, other may not, but our choice in our results format should not force our hand either way)

And to add to that, I think given a router that does that multiple rule evaluation then I think that could either write multiple results or write a single result that is an array. That's why I'm saying figuring out what our results look like is the important choice we want to make and settle on. How we populate those two different views of results is not likely to be a blocking point, though should obviously be thought out.

As an aside, looks like Excel does have select many support via things that look like `={"pizza", "burgers"}`. Google sheets strangely doesn't support this in the same way that I can tell.

**Comment by rowanseymour** [comment]
I don't see this is just a discussion over how we populate something. To me storing `1, 2, 4` as a single multi-value result asserts something about the state of `1, 2, 4` together (in this case they're a subset of a set of choices). 

A history of values a result has ever taken, doesn't come with the same assertion. Different nodes can append to the same result history with very different rules. There are potentially junk answers in their that lead to prompting for new input. It becomes a lot harder to get the set of values that together actually answered the poll question.

Re not always requiring all tokens match a rule, you could extend the idea of the `All/None` router to include `Some`:

![Captura de pantalla 2019-03-12 a la(s) 10 14 07](https://user-images.githubusercontent.com/675558/54211738-d3822c80-44af-11e9-943b-3583caf14d14.png)

On various sprints through this node it might (over)write a result like:

```json
{
  "name": "Fav Foods",
  "category": ["Other", "Other", "Other"],
  "value": [4, 5, 6]
}
```
&darr;
```json
{
  "name": "Fav Foods",
  "category": ["Pizza", "Burgers", "Other"],
  "value": [1, 2, 4]
}
```
&darr;
```json
{
  "name": "Fav Foods",
  "category": ["Pizza", "Burgers", "Sushi"],
  "value": [1, 2, 3]
}
```

Maybe you also add the concept of an aggregate category?

```json
{
  "name": "Fav Foods",
  "category": ["Pizza", "Burgers", "Sushi"],
  "aggregate_category": "All",
  "value": [1, 2, 3]
}
```

**Comment by nicpottier** [comment]
Ok, so lets explore array result values as you have.

Are categories always an array then or only sometimes? Seems it would be nice for that to be consistent either way.
How does that look in the context?
Same with values, are they always arrays or just sometimes? 
Do we allow types of any array? (seems we have to) 
Can arrays be of mixed types? (seem we have to so previous question is irrelevant) 
Do we add array types to contact fields? 
If not, what happens during assignment?

**Comment by rowanseymour** [comment]
So the simplest and dumbest approach would be to just model with CSV. Arrays already stringify to CSV so it's not as ridiculous as it might sound, and if you're tokenizing by things like commas then you can know that your values and categories aren't going to contain commas themselves. Where this falls down is on membership tests. `"urge" in "pizza, burgers, sushi"`.

Better approach is probably to say that values and categories are arrays of strings. Results values and categories already have to be strings. I don't think this is introducing types to results, I think it's just saying those things are multiple strings (have updated examples above). And I agree that it would be nice for all results to have the same schema.

Few different ways these might function in expressions..

**1. Distinct properties for single value vs multi-value access**

Storage-wise you have `.values` and `.categories` which are arrays. In expressions you have `.values[0]` `.categories[0]` or `.value` and `.category` which shortcut to the former.

**2. Leaning on auto-stringing of lists as CSV**

Currently `["x", "y"]` stringifies as `x, y` but `["x"]` stringifies as just `x`. So maybe `.value` and `.category` could actually be arrays of strings but they're usually going to look like strings. I don't like this as much because there will be places where they don't behave like just strings, e.g. `json(results.age.category)` &rarr; `["Old"]`

**3. Distinct properties for stringified vs array access**

Storage-wise you have `.values` and `.categories` which are arrays. In expressions you have `.values` `.categories` which are indexable arrays or `.value` and `.category` which are stringified (i.e. csv) versions of the former.

I'd lean toward 1.

Re contact fields, currently if you tried to assign `@(array("x", "y", "z"))` to a contact field of type TEXT you'd get `x, y, z`, and I think that would continue to be the case for arrays of result categories etc.

**Comment by nicpottier** [comment]
I like 1.. basically as you said saying we move to everything being plural but you have convenience shortcuts for single case. It's a bit weird to tell the user, hey `value` is just the *first* value, that makes sense right? But I do think it satisfies the keep things simple for the simple case and remove the veil of simplicity only when they need it.

What do we do in our API? Same thing? (both singular and plural?)

**Comment by rowanseymour** [comment]
After discussion with @nicpottier and @ericnewcomer, we're thinking about whether categories need to be moved to the router so that in the case of multiple choice, the router can accumulate the set of matching categories.

Currently a typical wait-route-split node looks like:

```json
{
	"uuid": "10e483a8-5ffb-4c4f-917b-d43ce86c1d65",
        "actions": [],
	"wait": {
		"type": "msg"
	},
	"router": {
		"type": "switch",
		"result_name": "Beer",
		"default_exit_uuid": "1fc4c133-d038-4f75-a69e-6e7e3190e5d8",
		"operand": "@input",
		"cases": [{
				"uuid": "d2f852ec-7b4e-457f-ae7f-f8b243c49ff5",
				"type": "has_any_word",
				"arguments": [
					"yeah yes"
				],
				"exit_uuid": "ea304225-332e-49d4-9768-1e804cd0b6c2"
			},
			{
				"uuid": "692926ea-09d6-4942-bd38-d266ec8d3716",
				"type": "has_any_word",
				"arguments": [
					"affirmative"
				],
				"exit_uuid": "ea304225-332e-49d4-9768-1e804cd0b6c2"
			}
		]
	},
	"exits": [{
			"uuid": "ea304225-332e-49d4-9768-1e804cd0b6c2",
			"destination_node_uuid": "5b977652-91e3-48be-8e86-7c8094b4aa8f",
			"name": "Yes"
		},
		{
			"uuid": "1fc4c133-d038-4f75-a69e-6e7e3190e5d8",
			"destination_node_uuid": "833fc698-d590-42dc-93e1-39e701b7e8e4",
			"name": "Other"
		}
	]
}
```

But we'd be introducing a categories layer between cases and exits, something like:

```json
{
	"uuid": "10e483a8-5ffb-4c4f-917b-d43ce86c1d65",
	"actions": [],
	"wait": {
		"type": "msg"
	},
	"router": {
		"type": "switch",
		"result_name": "Beer",
		"default_category_uuid": "1fc4c133-d038-4f75-a69e-6e7e3190e5d8",
		"operand": "@input",
		"cases": [{
				"uuid": "d2f852ec-7b4e-457f-ae7f-f8b243c49ff5",
				"type": "has_any_word",
				"arguments": [
					"yeah yes"
				],
				"category_uuid": "ea304225-332e-49d4-9768-1e804cd0b6c2"
			},
			{
				"uuid": "692926ea-09d6-4942-bd38-d266ec8d3716",
				"type": "has_any_word",
				"arguments": [
					"affirmative"
				],
				"category_uuid": "ea304225-332e-49d4-9768-1e804cd0b6c2"
			}
		],
		"categories": [{
				"uuid": "ea304225-332e-49d4-9768-1e804cd0b6c2",
				"exit_uuid": "5bd06585-7112-48a6-9506-bb15440a492f",
				"name": "Yes"
			},
			{
				"uuid": "1fc4c133-d038-4f75-a69e-6e7e3190e5d8",
				"exit_uuid": "c4aa3b73-9ebe-4cb1-8268-2df8e7ec274f",
				"name": "Other"
			}
		]
	},
	"exits": [{
			"uuid": "5bd06585-7112-48a6-9506-bb15440a492f",
			"destination_node_uuid": "5b977652-91e3-48be-8e86-7c8094b4aa8f"
		},
		{
			"uuid": "c4aa3b73-9ebe-4cb1-8268-2df8e7ec274f",
			"destination_node_uuid": "833fc698-d590-42dc-93e1-39e701b7e8e4"
		}
	]
}
```

**Comment by rowanseymour** [comment]
A further thought on result storage and exposure in expressions. `@result.foods.value` isn't a particularly self-explanatory variable name, so perhaps something like `match` or `text` would be better. Parallel arrays are also not always the best thing, especially with `category_localized` included as a 3rd. So maybe a result looks like:

```json
{
  "input": "I like burgers and pizza!",
  "values": [
    {"text": "burgers", "category": "Burgers", "category_localized": "Hamburguesas"},
    {"text": "pizza", "category": "Pizza"}
  ],
  "name" : "Foods"
}
```

Not using parallel arrays means:

- You can have category-less values from `set_run_result` actions
- Can drop `category_localized` when it isn't defined

And this would be exposed in expressions like:

`@result.foods` &rarr; `burgers, pizza` (CSV of each text, usually a single text)
`@result.foods.text` is same
`@result.foods.category` &rarr; `Burgers, Pizza` (CSV of each category, usually a single text)
`@(result.foods.values[0].text)` &rarr; `burgers`
`@(result.foods.values[0].category)` &rarr; `Pizza`
`@(result.foods.values[1].text)` &rarr; `pizza`
`@(length(result.foods.values))` &rarr; `2`

- For all the users who are never going to touch multiple choice, nothing changes
- It's kinda similar to how `@input` is the newline-separated concatenation of msg text and attachments, but you can also access attachments as an array with `@input.attachments`
- Clearly once you start unpacking `.values[..]` the expressions get a bit more complex but this is not going to be a common case.

The runs API format would be updated to include `.values` but wouldn't conflict with `.value` or `.category`

**Comment by nicpottier** [comment]
Hrmm, agree lots of // arrays starts getting weird, what about:

```json
{
  "input": "I like burgers and pizza!",
  "matches": [
    {"value": "burgers", "category": "Burgers", "category_localized": "Hamburguesas"},
    {"value": "pizza", "category": "Pizza"}
  ],
  "name" : "Foods"
}
```

Only because `matches` seems to map well to these are tests and there is one `match` per case that, well, matched.

I do feel that `value` is less ambiguous than `text` only because `text` and `input` starts getting a bit goofy especially when talking about text message inputs.

The shortcut stuff you have above kinda matches the tom-foolery we do with `@contact.urn` so I guess I'm ok with that.

**Comment by rowanseymour** [comment]
We may have agreed on this for results storage:

```json
{
    "input": "I like burgers and pizza!",
    "values": [
        {"value": "burgers", "category": "Burgers", "localized_category": "Hamburguesas"},
        {"value": "pizza", "category": "Pizza"}
    ],
    "name": "Foods"
}
```

and expressions like:

`@result.foods` &rarr; `burgers` (`.value` of first value)
`@result.foods.value` is same
`@result.foods.category` &rarr; `Burgers` (`.category` of first value)
`@result.foods.localized_category` &rarr; `Hamburguesas` (`.localized_category` of first value)

`@result.foods.values` &rarr; `[burgers, pizza]` (array of strings)
`@result.foods.categories` &rarr; `[Burgers, Pizza]` (array of strings)
`@result.foods.localized_categories` &rarr; `[Hamburguesas, Pizza]` (array of strings)

NB. I switched `category_localized` to `localized_category` for good measure...
